### ISSUE TITLE
Add `AnimationTheme` to specify default curve(s) and duration(s)

### ISSUE THREAD

**Issue by caseycrogers** [flutter/flutter#155573]
### Use case

There are a ton of super useful [animated widgets](https://api.flutter.dev/flutter/widgets/AnimatedWidget-class.html). Most require an explicit duration argument and/or Curve argument. Where defaults are available, they're handled on a case by case basis.

The most obvious way to handle these arguments across your app it to create globals and/or create your own inherited widget to hold them. This works okay, but it's especially annoying in the following cases:

It's annoying to have to specify the relevant curve(s) and duration(s) in every relevant place. It's even more annoying if you:
1. Use different durations and/or curves for different scenarios
2. Want to dynamically override the animation duration and/or curve according to user settings or situation (ie tuning a game to feel faster and snappier when a user is on a streak)
3. Are on a large team with specific animation design guidelines and don't want to rely on developers remembering to pipe the correct values into the relevant widgets

This is an especially good candidate for including into the framework itself instead of relying on the community to build packages/roll their own solutions because the core value add (having sane defaults and optional arguments on all relevant built-ins) can only be achieved by changing the built in widgets.

### Proposal

1. Create an `AnimationTheme` inherited widget and  `AnimationThemeData` class that hold a default `Curve` and `Duration`
2. Add an `AnimationTheme.of(context)` that retrieves the animation theme and listens to any changes in the theme
2. Add `AnimationThemeData` to `ThemeData` and include default values
3. Have all implicitly animated widgets inherit default values (with chained inheritance resolution if `inherit=true`)  from the nearest `AnimationTheme`/`Theme`
4. Make the `duration` and `curve` arguments optional and use them as overrides for the theme values when non-null

Open questions:
1. Should there be other arguments such as `curveFast`, `durationFast` etc equivalent to `bodySmall`, `bodyMedium`....? Devs could then opt-into alternate theme-held animations
2. Are there other widgets that should reference the theme held values? Maybe navigator and it's default transitions?
3. Should we lerp curves and durations in some clever way (like `DefaultTextStyle` etc) when their theme values change, or should we defer to the individual widgets to handle changes as they see fit?

On all of the above I feel the best approach is to launch an MVP and then see what feature requests devs make as all these extra features are easy to add later if needed but hard to roll back if we find they're not.

**Comment by dkwingsmt** [comment]
I agree that it definitely feels awkward and inconvenient to not include animation values in themes!

cc @MitchellGoodwin for theming API

I'm not sure if a new class `AnimationTheme` is sufficient though. I can see why we want to group animatino-related settings in a separate class, but instead of just one global theme we might want to also include one instance for each widget that animates.

Also note: If we work on this we need to make sure that whatever API we come up with also works for `Simulation` animation as well.

**Comment by caseycrogers** [comment]
While it has advantages, I think a theme for each relevant widget will be far too many themes to be practical-and generally Flutter seems to have moved away from Widget specific themes to global semantic themes.

I suppose a layered approach could work here-have a global semantic theme and then have widget specific themes that inherit from the global theme. In which case, we'd still want to have the global semantic theme, so I vote we start with just the most barebones theme proposed here and then add more theming options in the future if it makes sense? This should be a good approach as I think we'll want this basic theme no matter what approach we take down the line-it's easy to add but very hard to remove features so lets start broad and simple.

I'm happy to write up a PR for this, but would love some input first from folks at Google to make sure we're on the right track wrt top level design.

**Comment by dkwingsmt** [comment]
Yeah, we can always start with the global theme. And I agree with your idea that a starting MVP should be important to determining the design. The answer to your open questions should rely on how animation customization is needed by real projects. Did you come up with this proposal because it was needed in your projects?

**Comment by Piinks** [comment]
Something to note here, the theming does not current exist outside of the material or cupertino library. This implies we would add the concept of theming to the widgets layer for the first time. We should be super considerate about what we add and how, since it could set the example for how we will support generic theming outside of the design language libraries.

**Comment by Piinks** [comment]
Actually.. thinking on this more.. What if we added something similar to ScrollConfiguration/ScrollBehavior, a la AnimationConfiguration/AnimationBehavior? It is inherited like theming, and would avoid the complexity of defining a the generic theming strategy for the widgets libary.

**Comment by caseycrogers** [comment]
@dkwingsmt I publish and manage an "app like" game built in Flutter that is pretty animation heavy so this need comes from my app directly-though I have also heard it from other app developers in their apps. Later this week I'll try to run through my app as a case study and document all the ways it is using animations and how I'm handling this today (globally declared `kDurationFast`, `kDuration`, etc and ad-hoc curves).

@Piinks Ah I see-yeah I think defining this independent of the Theme system makes sense then. From looking at ScrollBehavior, it looks like it behaves pretty similarly to the theme system, does that sound right to you? How about I write up a detailed design proposal and we can review it-it feels like there might be enough sticky details to justify a doc over just a Github issue.

**Comment by Piinks** [comment]
SGTM! I'd be happy to read it and share around for feedback!

**Comment by dkwingsmt** [comment]
> Later this week I'll try to run through my app as a case study and document all the ways it is using animations and how I'm handling this today

Thank you, yes that would be very helpful. 

> What if we added something similar to ScrollConfiguration/ScrollBehavior, a la AnimationConfiguration/AnimationBehavior?

Yeah, I agree. In general making something globally configurable only requires an inherited widget, which each widget can already do by themselves. Adding another field to `ThemeData` is another topic and needs stronger reasoning, such as "this is a property applicable to almost all widgets" - which _can_ be very much true for animation properties.

**Comment by MitchellGoodwin** [comment]
+1 to a non-theme inherited widget.

If you did want to do it through the theme, I'd be curious if Materials theme extensions are able to handle it well, or if there are issues with equality checks causing rebuilds.

**Comment by caseycrogers** [comment]
Sounds like there's consensus on non-theme, all that makes sense to me!

Haven't gotten to writing up a design doc yet, hopefully will get to it soon. In the meantime, here's a case study I just wrote up for [dribblegame.com](https://dribblegame.com/). The case study isn't strictly tied to this issue, it's a provide overview that covers A LOT. The repo is unfortunately private but if any of you all would like to see the source code I can add you as a read only collaborator and/or do a video call to walk you through the source code. There is A LOT of animation logic in Dribble Game and the referenced doc only covers a (fairly representative) sample of it.

[Dribble Game Animation Case Study](https://docs.google.com/document/d/14ZKbMXq3del-wegsxL7xtLGfV6yMipFBCN_Pb2xVoac/edit?usp=sharing)

**Comment by Piinks** [comment]
Thinking on this some more...

> Proposal
> 1. Create an AnimationTheme inherited widget and AnimationThemeData class that hold a default Curve and Duration
> 2. Add an AnimationTheme.of(context) static method that retrieves the animation theme and listens to any changes in the theme
> 3. Add AnimationThemeData to ThemeData and include default values
> 4. Have all implicitly animated widgets inherit default values (with chained inheritance resolution if inherit=true) from the nearest AnimationTheme/Theme
> 5. Make the duration and curve arguments optional and nullable with no default values and use them as overrides for the theme values when non-null

Some questions to include in the doc - there are so many animations that could be running through an app, would this affect route transition animations? Built-in animations like when a FAB changes position or scrollbar visibility animations for example? And how would a given animated widget opt out of inheriting this configuration? Lots of animated widgets have differing default behaviors, curves, durations etc. Being able to set a clear scope for the inherited animation behaviors, without inadvertently affecting other unintended animations will be interesting to solve.

**Comment by caseycrogers** [comment]
Here's my first pass at a design doc:
[https://docs.google.com/document/d/18c28hbe1mRBZLcnaip5LM9zlshIOEpuvBLWKZiXGe9g/edit?usp=sharing](Inherited Defaults for Widget Animations)

Comments welcome! It'll likely need a decent bit more input and editing before it's in green-lightable shape.

**Comment by caseycrogers** [comment]
I had some helpful back and forth with @nate-thegrate in the [Design Doc](https://docs.google.com/document/d/18c28hbe1mRBZLcnaip5LM9zlshIOEpuvBLWKZiXGe9g/edit) and refined the proposal a good bit.

Biggest insight is that there's already `AnimatedStyle` used by `MaterialApp` and a few inheriting widgets for us to piggyback this proposal off of (details in doc). I think it's at a point where I'll start working on a PR, but I'm about to move and start a new job so I may have to put this on ice for a bit and come back to it. Still open to any feedback in the associated doc, and in the upcoming PR!
